hail/src/main/scala/is/hail/expr/ir/agg/StagedBlockLinkedList.scala

```diff
@@ -0,0 +1,242 @@
+      tmpNode := r.allocate(nodeType.alignment, nodeType.byteSize),
+      initNode(tmpNode,
+        buf = r.allocate(bufferType.contentsAlignment, bufferType.contentsByteSize(cap)),
+        count = 0),
+      bufferType.stagedInitialize(buffer(tmpNode), cap))
+    (setup, tmpNode)
+  }
+
+  def initWithCapacity(r: Code[Region], initialCap: Code[Int]): Code[Unit] = {
+    val (setup, n) = allocateNode(r, initialCap)
+    Code(setup,
+      firstNode := n,
+      lastNode := n,
+      totalCount := 0)
+  }
+
+  def init(r: Code[Region]): Code[Unit] =
+    initWithCapacity(r, defaultBlockCap)
+
+  private[agg] def pushNewBlockNode(r: Code[Region], cap: Code[Int]): Code[Unit] = {
+    val (setup, newNode) = allocateNode(r, cap)
+    Code(setup,
+      setNext(lastNode, newNode),
+      lastNode := newNode)
+  }
+
+  private[agg] def foreachNode(f: Node => Code[Unit]): Code[Unit] = {
+    Code(
+      tmpNode := firstNode,
+      p := true,
+      Code.whileLoop(p,
+        f(tmpNode),
+        p := hasNext(tmpNode),
+        tmpNode := next(tmpNode)))
+  }
+
+  def foreach(f: EmitTriplet => Code[Unit]): Code[Unit] = {
+    def et(n: Node, i: Code[Int]) =
+      EmitTriplet(Code._empty,
+        bufferType.isElementMissing(buffer(n), i),
+        Region.loadIRIntermediate(elemType)(
+          bufferType.elementOffset(buffer(n), capacity(n), i)))
+    foreachNode { n => Code(
+      i := 0,
+      Code.whileLoop(i < count(n),
+        f(et(n, i)),
+        i := i + 1))
+    }
+  }
+
+  def push(r: Code[Region], elt: EmitTriplet): Code[Unit] = {
+    Code(
+      (count(lastNode) >= capacity(lastNode)).orEmpty(
+        pushNewBlockNode(r, defaultBlockCap)), // push a new block if lastNode is full
+      elt.setup,
+      elt.m.mux(
+        pushMissing(lastNode),
+        pushPresent(lastNode, { dst =>
+          if(elemType.isPrimitive)
+            Region.storePrimitive(elemType, dst)(elt.value)
+          else
+            StagedRegionValueBuilder.deepCopy(fb, r, elemType, elt.value, dst)
+        })),
+      totalCount := totalCount + 1)
+  }
+
+  def append(r: Code[Region], bll: StagedBlockLinkedList): Code[Unit] = {
+    assert(bll.elemType.isOfType(elemType))
+    bll.foreach(push(r, _))
+  }
+
+  def writeToSRVB(srvb: StagedRegionValueBuilder): Code[Unit] = {
+    assert(srvb.typ.isOfType(bufferType))
+    Code(
+      srvb.start(totalCount, init = true),
+      foreach { elt =>
+        Code(
+          elt.m.mux(
+            srvb.setMissing(),
+            srvb.addWithDeepCopy(elemType, elt.value)),
+          srvb.advance())
+      })
+  }
+
+  def serialize(r: Code[Region], ob: Code[OutputBuffer]): Code[Unit] = {
+    Code(
+      foreachNode { n =>
+        // NOTE: we are only copying a portion of the missingness bytes, only as many as there are
+        // present (specifically: 'count(n)', not 'capacity(n)'). we can't use emitArray because it
+        // would mistakenly copy elements according to the capacity ...
```

Comment:
What about adding an optional parameter to `emitArray` to allow serializing a prefix of an array? That avoids duplicating serialization logic here.